modules/auxiliary/scanner/http/cisco_secure_desktop.rb

```diff
@@ -0,0 +1,337 @@
+            'uri' => '/+CSCOE+/logon.html',
+            'method' => 'GET',
+            'headers' => headers,
+          )
+
+    if res &&
+       res.code == 302
+
+      res = send_request_raw(
+              'uri' => '/CACHE/sdesktop/install/start.htm',
+              'method' => 'GET',
+              'headers' => headers,
+            )
+    end
+
+    if res &&
+       res.code == 200 &&
+       res.body.match(/Cisco Secure Desktop/)
+
+      return true
+    else
+      return false
+    end
+  end
+
+  # Verify whether we're allowed to use password to login or not
+  def is_pwd_login_allowed?
+    headers = {
+      'Accept' => 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
+      'Accept-Language' => 'en-US,en;q=0.5',
+      'Cookie' => "webvpnlogin=1; webvpnLang=en; sdesktop=#{$sdesktop_cookie};",
+     }
+
+      res = send_request_raw(
+              'uri' => '/+CSCOE+/logon.html?fcadbadd=1',
+              'method' => 'GET',
+              'headers' => headers,
+            )
+
+    if res &&
+       res.code == 200 &&
+       res.body.match(/password_field/)
+
+      return true
+    else
+      return false
+    end
+  end
+
+  def do_logout(cookie)
+    res = send_request_raw(
+            'uri' => '/+webvpn+/webvpn_logout.html',
+            'method' => 'GET',
+            'headers' => {
+               'Accept' => 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
+               'Accept-Language' => 'en-US,en;q=0.5',
+               'Cookie' => "webvpnlogin=1; webvpnLang=en; sdesktop=#{$sdesktop_cookie};",
+              }
+          )
+  end
+
+  def report_cred(opts)
+    service_data = {
+      address: opts[:ip],
+      port: opts[:port],
+      service_name: 'Cisco Secure Desktop',
+      protocol: 'tcp',
+      workspace_id: myworkspace_id
+    }
+
+    credential_data = {
+      origin_type: :service,
+      module_fullname: fullname,
+      username: opts[:user],
+      private_data: opts[:password],
+      private_type: :password
+    }.merge(service_data)
+
+    login_data = {
+      last_attempted_at: DateTime.now,
+      core: create_credential(credential_data),
+      status: Metasploit::Model::Login::Status::SUCCESSFUL,
+      proof: opts[:proof]
+    }.merge(service_data)
+
+    create_credential_login(login_data)
+  end
+
+  def get_token_cookie()
+    vprint_status("Requesting a session token...")
+    res = send_request_raw(
+            'uri' => '/+CSCOE+/sdesktop/token.xml',
+            'method' => 'GET',
+            'headers' => {
+               'Accept' => 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
+               'Accept-Language' => 'en-US,en;q=0.5',
+               'Referer' => "https://#{rhost}:#{rport}/CACHE/sdesktop/install/start.htm",
+               'Cookie' => 'webvpnlogin=1; webvpnLang=en',
+              }
+          )
+
+    if res &&
+       res.code == 200
+
+      token = res.body.scan(/<token>([A-Z0-9]{24})/).flatten[0]
```

Comment:
Not a big deal - `flatten.first` is more aesthetically pleasing.

Reply:
Seems to work; will commit the change.